kernel/msg_q.c

```diff
@@ -195,6 +197,83 @@ static inline int z_vrfy_k_msgq_put(struct k_msgq *msgq, const void *data,
 #include <syscalls/k_msgq_put_mrsh.c>
 #endif /* CONFIG_USERSPACE */
 
+int z_impl_k_msgq_put_front(struct k_msgq *msgq, const void *data, k_timeout_t timeout)
+{
+	__ASSERT(!arch_is_in_isr() || K_TIMEOUT_EQ(timeout, K_NO_WAIT), "");
+
+	struct k_thread *pending_thread;
+	k_spinlock_key_t key;
+	int result;
+
+	key = k_spin_lock(&msgq->lock);
+
+	SYS_PORT_TRACING_OBJ_FUNC_ENTER(k_msgq, put, msgq, timeout);
+
+	if (msgq->used_msgs < msgq->max_msgs) {
+		/* message queue isn't full */
+		pending_thread = z_unpend_first_thread(&msgq->wait_q);
+		if (pending_thread != NULL) {
+			SYS_PORT_TRACING_OBJ_FUNC_EXIT(k_msgq, put, msgq, timeout, 0);
+
+			/* give message to waiting thread */
+			(void)memcpy(pending_thread->base.swap_data, data,
+			       msgq->msg_size);
+			/* wake up waiting thread */
+			arch_thread_return_value_set(pending_thread, 0);
+			z_ready_thread(pending_thread);
+			z_reschedule(&msgq->lock, key);
+			result = 0;
+			goto out;
+		} else {
+			/* put message in queue */
+			/* Only different with z_impl_k_msgq_put */
+			__ASSERT_NO_MSG(msgq->read_ptr >= msgq->buffer_start &&
+					msgq->read_ptr < msgq->buffer_end);
+			if (msgq->read_ptr == msgq->buffer_start) {
+				msgq->read_ptr = msgq->buffer_end;
+			}
+			msgq->read_ptr -= msgq->msg_size;
+			(void)memcpy(msgq->read_ptr, data, msgq->msg_size);
+			msgq->used_msgs++;
+#ifdef CONFIG_POLL
+			handle_poll_events(msgq, K_POLL_STATE_MSGQ_DATA_AVAILABLE);
+#endif /* CONFIG_POLL */
+		}
+		result = 0;
+	} else if (K_TIMEOUT_EQ(timeout, K_NO_WAIT)) {
+		/* don't wait for message space to become available */
+		result = -ENOMSG;
+	} else {
+		SYS_PORT_TRACING_OBJ_FUNC_BLOCKING(k_msgq, put, msgq, timeout);
+
+		/* wait for put message success, failure, or timeout */
+		_current->base.swap_data = (void *) data;
+
+		result = z_pend_curr(&msgq->lock, key, &msgq->wait_q, timeout);
+		SYS_PORT_TRACING_OBJ_FUNC_EXIT(k_msgq, put, msgq, timeout, result);
+		return result;
```

Comment:
I think that allowing the caller to pend with a timeout can lead to inconsistent behaviour. That is, if the msgq buffer is not full, then the message will be copied into the front of the buffer. However, if the msgq buffer is full, then the message is being copied into the end of the buffer when we get a message from the msgq (and subsequently remove it from the wait queue).  Because of this, I am wondering if the timeout parameter ought to be removed completely, and k_msgq_put_front() be treated as if it had a timeout of K_NO_WAIT.

Reply:
> I think that allowing the caller to pend with a timeout can lead to inconsistent behaviour. 

Sorry, I don't get the point.

> That is, if the msgq buffer is not full, then the message will be copied into the front of the buffer. 
However, if the msgq buffer is full, then the message is being copied into the end of the buffer when we get a message from the msgq (and subsequently remove it from the wait queue). 

The behavior is correct.

> Because of this, I am wondering if the timeout parameter ought to be removed completely, and k_msgq_put_front() be treated as if it had a timeout of K_NO_WAIT.

I don't get the point, too. Could you explain more detail?

Reply:
There is an implicit property being associated with the message--the message should take priority and be placed at the front of the message queue. As long as the message queue buffer is not full (and thus there is space to put this prioritized message) this works correctly.

My concern is that when the message queue buffer is full and the caller has to pend, then that implicit property is lost and the caller has no indication that this happened or could happen.

Imagine the following scenario ...

The message queue buffer is full. Thread A calls k_msgq_put_front(), but since the buffer is full, Thread A pends. Thread B comes along and calls k_msgq_get(). It does not get message A; it gets message X--the first in the message queue buffer. Internally, k_msgq_get() will then unpend the first waiting thread (which could be Thread A) and copy its message to the end of the message queue buffer. The implicit property of message A is lost, and there is no indication to the caller of k_msgq_put_front() that the message was copied to an unexpected location.

Reply:
> Imagine the following scenario ...
> 
> The message queue buffer is full. Thread A calls k_msgq_put_front(), but since the buffer is full, Thread A pends. Thread B comes along and calls k_msgq_get(). It does not get message A; it gets message X--the first in the message queue buffer. Internally, k_msgq_get() will then unpend the first waiting thread (which could be Thread A) and copy its message to the end of the message queue buffer. The implicit property of message A is lost, and there is no indication to the caller of k_msgq_put_front() that the message was copied to an unexpected location.

You are right!!!!
When `thread A` wake up, it will copy data to `write_ptr` but In my expectation, it should move `read_ptr` and then copy.
The main reason is we don't record the reason make `thread A` sleep.
We should record the pending reason is `k_msgq_put` or `k_msgq_put_front`

But  there is no bit can be used in `thread_state`.
Do you have any suggestion?

Reply:
> There is an implicit property being associated with the message--the message should take priority and be placed at the front of the message queue. As long as the message queue buffer is not full (and thus there is space to put this prioritized message) this works correctly.
> 
> My concern is that when the message queue buffer is full and the caller has to pend, then that implicit property is lost and the caller has no indication that this happened or could happen.
> 
> Imagine the following scenario ...
> 
> The message queue buffer is full. Thread A calls k_msgq_put_front(), but since the buffer is full, Thread A pends. Thread B comes along and calls k_msgq_get(). It does not get message A; it gets message X--the first in the message queue buffer. Internally, k_msgq_get() will then unpend the first waiting thread (which could be Thread A) and copy its message to the end of the message queue buffer. The implicit property of message A is lost, and there is no indication to the caller of k_msgq_put_front() that the message was copied to an unexpected location.

You are right!!!!

When thread A wake up, it will copy data to write_ptr but In my expectation, it should move read_ptr and then copy.

The main reason is we don't record the reason make thread A sleep.
We should record the pending reason is k_msgq_put or k_msgq_put_front and when thread A unpend, it have to execute the corresponding operation.

But there is no bit can be used in thread_state.
Do you have any suggestion?